File: app/src/main/java/org/runnerup/core/util/ActivitySummaryBinder.java
package org.runnerup.core.util;

import android.view.View;
import android.widget.TextView;

/** Binds distance, elapsed time, and pace/speed for activity list rows. */
public final class ActivitySummaryBinder {

  private ActivitySummaryBinder() {}

  public static void bind(
      Formatter formatter,
      TextView distanceView,
      TextView timeView,
      TextView paceView,
      double distanceMeters,
      long durationSeconds) {
    bind(
        formatter,
        distanceView,
        timeView,
        paceView,
        Formatter.Format.TXT_SHORT,
        Formatter.Format.TXT_SHORT,
        distanceMeters,
        durationSeconds);
  }

  public static void bind(
      Formatter formatter,
      TextView distanceView,
      TextView timeView,
      TextView paceView,
      Formatter.Format distanceFormat,
      Formatter.Format paceFormat,
      double distanceMeters,
      long durationSeconds) {
    distanceView.setText(
        formatter.formatDistance(distanceFormat, Math.round(distanceMeters)));
    timeView.setText(formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds));
    if (durationSeconds > 0) {
      paceView.setText(
          formatter.formatVelocityByPreferredUnit(
              paceFormat, distanceMeters / durationSeconds));
      paceView.setVisibility(View.VISIBLE);
    } else {
      paceView.setText("");
    }
  }

  /** Activity detail header: hides pace row when duration is zero. */
  public static void bindActivityHeader(
      Formatter formatter,
      TextView distanceView,
      TextView timeView,
      TextView paceView,
      View paceSeparator,
      double distanceMeters,
      long durationSeconds) {
    distanceView.setText(
        formatter.formatDistance(Formatter.Format.TXT_SHORT, Math.round(distanceMeters)));
    if (durationSeconds > 0) {
      timeView.setText(
          formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds));
      paceView.setVisibility(View.VISIBLE);
      if (paceSeparator != null) {
        paceSeparator.setVisibility(View.VISIBLE);
      }
      paceView.setText(
          formatter.formatVelocityByPreferredUnit(
              Formatter.Format.TXT_LONG, distanceMeters / durationSeconds));
    } else {
      timeView.setText(
          durationSeconds == 0 && distanceMeters > 0
              ? ""
              : formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds));
      paceView.setVisibility(View.GONE);
      if (paceSeparator != null) {
        paceSeparator.setVisibility(View.GONE);
      }
    }
  }
}

File: app/src/main/kotlin/org/runnerup/core/util/ActivitySummaryBinder.kt
package org.runnerup.core.util

import android.view.View
import android.widget.TextView

/** Binds distance, elapsed time, and pace/speed for activity list rows. */
object ActivitySummaryBinder {
  @JvmStatic
  fun bind(
      formatter: Formatter,
      distanceView: TextView,
      timeView: TextView,
      paceView: TextView,
      distanceMeters: Double,
      durationSeconds: Long,
  ) {
    bind(
        formatter,
        distanceView,
        timeView,
        paceView,
        Formatter.Format.TXT_SHORT,
        Formatter.Format.TXT_SHORT,
        distanceMeters,
        durationSeconds,
    )
  }

  @JvmStatic
  fun bind(
      formatter: Formatter,
      distanceView: TextView,
      timeView: TextView,
      paceView: TextView,
      distanceFormat: Formatter.Format,
      paceFormat: Formatter.Format,
      distanceMeters: Double,
      durationSeconds: Long,
  ) {
    distanceView.text =
        formatter.formatDistance(distanceFormat, kotlin.math.round(distanceMeters).toLong())
    timeView.text = formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds)
    if (durationSeconds > 0) {
      paceView.text =
          formatter.formatVelocityByPreferredUnit(
              paceFormat,
              distanceMeters / durationSeconds,
          )
      paceView.visibility = View.VISIBLE
    } else {
      paceView.text = ""
    }
  }

  @JvmStatic
  fun bindActivityHeader(
      formatter: Formatter,
      distanceView: TextView,
      timeView: TextView,
      paceView: TextView,
      paceSeparator: View?,
      distanceMeters: Double,
      durationSeconds: Long,
  ) {
    distanceView.text =
        formatter.formatDistance(Formatter.Format.TXT_SHORT, kotlin.math.round(distanceMeters).toLong())
    if (durationSeconds > 0) {
      timeView.text = formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds)
      paceView.visibility = View.VISIBLE
      paceSeparator?.visibility = View.VISIBLE
      paceView.text =
          formatter.formatVelocityByPreferredUnit(
              Formatter.Format.TXT_LONG,
              distanceMeters / durationSeconds,
          )
    } else {
      timeView.text =
          if (durationSeconds == 0L && distanceMeters > 0) {
            ""
          } else {
            formatter.formatElapsedTime(Formatter.Format.TXT_SHORT, durationSeconds)
          }
      paceView.visibility = View.GONE
      paceSeparator?.visibility = View.GONE
    }
  }
}
